File: mf2/java/src/main/java/com/box/l10n/mojito/mf2/TranslateDemo.java
package com.box.l10n.mojito.mf2;

import java.nio.file.Path;
import java.util.Map;

public final class TranslateDemo {
    private TranslateDemo() {}

    public static void main(String[] args) throws Exception {
        Catalog catalog = Catalog.load(Path.of("../examples/catalog.json"), DemoFunctions.registry());
        System.out.println("welcome[fr] -> \"" + catalog.translate("welcome", "fr", Map.of("name", "Mojito")) + "\"");
        System.out.println("welcome[fr-CA] -> \"" + catalog.translate("welcome", "fr-CA", Map.of("name", "Mojito")) + "\"");
        System.out.println("checkout.total[en] -> \""
                + catalog.translate("checkout.total", "en", Map.of("amount", 1234.5)) + "\"");
        System.out.println("checkout.total[fr] -> \""
                + catalog.translate("checkout.total", "fr", Map.of("amount", 1234.5)) + "\"");
        System.out.println("cart.items[en] -> \"" + catalog.translate("cart.items", "en", Map.of("count", 1)) + "\"");
        System.out.println("cart.items[en] -> \"" + catalog.translate("cart.items", "en", Map.of("count", 5)) + "\"");
        System.out.println("cart.items[ru] -> \"" + catalog.translate("cart.items", "ru", Map.of("count", 2)) + "\"");
        System.out.println("cart.items[ru] -> \"" + catalog.translate("cart.items", "ru", Map.of("count", 5)) + "\"");
    }

    private record Catalog(Map<String, Object> messages, Mf2FunctionRegistry functions) {
        static Catalog load(Path path, Mf2FunctionRegistry functions) throws Exception {
            return new Catalog(object(object(JsonParser.parse(path)).get("messages")), functions);
        }

        String translate(String id, String locale, Map<String, ?> arguments) throws Mf2Exception {
            Map<String, Object> localized = object(messages.get(id));
            Object model = LocaleKey.lookup(localized, locale);
            return Mf2Message.fromJson(model).format(arguments, locale, functions);
        }
    }

    @SuppressWarnings("unchecked")
    private static Map<String, Object> object(Object value) {
        return (Map<String, Object>) value;
    }
}

File: mf2/rust/mf2-prototype/examples/translate_demo.rs
use std::collections::BTreeMap;
use std::fs;
use std::path::Path;

use demo_functions::demo_function_registry;
use mf2_prototype::{
    format_model_with_locale_and_functions_and_bidi, lookup_locale, BidiIsolation,
    FunctionRegistry, MessageModel,
};
use serde::Deserialize;

mod demo_functions;

#[derive(Debug, Deserialize)]
struct Catalog {
    messages: BTreeMap<String, BTreeMap<String, MessageModel>>,
}

impl Catalog {
    fn load(path: &Path) -> Self {
        let contents = fs::read_to_string(path)
            .unwrap_or_else(|error| panic!("failed to read {}: {error}", path.display()));
        serde_json::from_str(&contents)
            .unwrap_or_else(|error| panic!("failed to parse {}: {error}", path.display()))
    }

    fn translate(
        &self,
        message_id: &str,
        locale: &str,
        arguments: BTreeMap<String, serde_json::Value>,
        functions: &FunctionRegistry,
        bidi_isolation: BidiIsolation,
    ) -> String {
        let model = self.model(message_id, locale);
        format_model_with_locale_and_functions_and_bidi(
            model,
            &arguments,
            locale,
            functions,
            bidi_isolation,
        )
        .unwrap_or_else(|diagnostic| panic!("format failed: {diagnostic:?}"))
    }

    fn model(&self, message_id: &str, locale: &str) -> &MessageModel {
        let localized = self
            .messages
            .get(message_id)
            .unwrap_or_else(|| panic!("missing message id: {message_id}"));
        lookup_locale(localized, locale, "en")
            .unwrap_or_else(|| panic!("missing locale {locale} and fallback en for {message_id}"))
    }
}

fn main() {
    let catalog = Catalog::load(Path::new("../../examples/catalog.json"));
    let functions = demo_function_registry();
    let examples = [
        (
            "welcome",
            "fr",
            args([("name", serde_json::Value::String("Mojito".to_string()))]),
            BidiIsolation::None,
            "Bienvenue, Mojito !",
        ),
        (
            "welcome",
            "fr-CA",
            args([("name", serde_json::Value::String("Mojito".to_string()))]),
            BidiIsolation::None,
            "Bienvenue, Mojito !",
        ),
        (
            "checkout.total",
            "en",
            args([("amount", serde_json::Value::from(1234.5))]),
            BidiIsolation::None,
            "Total: $1,234.50",
        ),
        (
            "checkout.total",
            "fr",
            args([("amount", serde_json::Value::from(1234.5))]),
            BidiIsolation::None,
            "Total : 1\u{202f}234,50 €",
        ),
        (
            "file.saved",
            "en",
            args([(
                "fileName",
                serde_json::Value::String("שלום.txt".to_string()),
            )]),
            BidiIsolation::Default,
            "File \u{2068}שלום.txt\u{2069} saved.",
        ),
        (
            "cart.items",
            "en",
            args([("count", 1.into())]),
            BidiIsolation::None,
            "1 item",
        ),
        (
            "cart.items",
            "en",
            args([("count", 5.into())]),
            BidiIsolation::None,
            "5 items",
        ),
        (
            "cart.items",
            "ru",
            args([("count", 2.into())]),
            BidiIsolation::None,
            "2 предмета",
        ),
        (
            "cart.items",
            "ru",
            args([("count", 5.into())]),
            BidiIsolation::None,
            "5 предметов",
        ),
        (
            "assignee.files",
            "en",
            args([
                ("gender", serde_json::Value::String("male".to_string())),
                ("count", 1.into()),
            ]),
            BidiIsolation::None,
            "He reviewed 1 file",
        ),
        (
            "assignee.files",
            "en",
            args([
                ("gender", serde_json::Value::String("female".to_string())),
                ("count", 3.into()),
            ]),
            BidiIsolation::None,
            "She reviewed 3 files",
        ),
        (
            "assignee.files",
            "en",
            args([
                ("gender", serde_json::Value::String("unknown".to_string())),
                ("count", 2.into()),
            ]),
            BidiIsolation::None,
            "They reviewed 2 files",
        ),
    ];

    for (message_id, locale, arguments, bidi_isolation, expected) in examples {
        let actual = catalog.translate(message_id, locale, arguments, &functions, bidi_isolation);
        assert_eq!(actual, expected, "{message_id}/{locale}");
        println!("{message_id}[{locale}] -> \"{actual}\"");
    }
}

fn args<const N: usize>(
    values: [(&'static str, serde_json::Value); N],
) -> BTreeMap<String, serde_json::Value> {
    values
        .into_iter()
        .map(|(key, value)| (key.to_string(), value))
        .collect()
}
